can/dbc.cc

```diff
@@ -1,27 +1,251 @@
+    } else if (startswith(line, "SG_ ")) {
+      // new signal
+      int offset = 0;
+      if (!std::regex_search(line, match, sg_regexp)) {
+        bool ret = std::regex_search(line, match, sgm_regexp);
+        DBC_ASSERT(ret, "bad SG " << line);
+        offset = 1;
+      }
+      Signal& sig = signals[address].emplace_back();
+      sig.name = match[1].str();
+      sig.start_bit = std::stoi(match[offset + 2].str());
+      sig.size = std::stoi(match[offset + 3].str());
+      sig.is_little_endian = std::stoi(match[offset + 4].str()) == 1;
+      sig.is_signed = match[offset + 5].str() == "-";
+      sig.factor = std::stod(match[offset + 6].str());
+      sig.offset = std::stod(match[offset + 7].str());
+      set_signal_type(sig, address, checksum.get(), dbc_name);
+      if (sig.is_little_endian) {
+        sig.lsb = sig.start_bit;
+        sig.msb = sig.start_bit + sig.size - 1;
+      } else {
+        auto it = find(be_bits.begin(), be_bits.end(), sig.start_bit);
+        sig.lsb = be_bits[(it - be_bits.begin()) + sig.size - 1];
+        sig.msb = sig.start_bit;
+      }
+      DBC_ASSERT(sig.lsb < (64 * 8) && sig.msb < (64 * 8), "Signal out of bounds : " << line);
+    } else if (startswith(line, "VAL_ ")) {
+      // new signal value/definition
+      bool ret = std::regex_search(line, match, val_regexp);
+      DBC_ASSERT(ret, "bad VAL " << line);
+
+      auto& val = dbc->vals.emplace_back();
+      val.address = std::stoul(match[1].str());  // could be hex
+      val.name = match[2].str();
+
+      auto defvals = match[3].str();
+      std::regex regex{R"([\"]+)"};  // split on "
+      std::sregex_token_iterator it{defvals.begin(), defvals.end(), regex, -1};
+      // convert strings to UPPER_CASE_WITH_UNDERSCORES
+      std::vector<std::string> words{it, {}};
+      for (auto& w : words) {
+        w = trim(w);
+        std::transform(w.begin(), w.end(), w.begin(), ::toupper);
+        std::replace(w.begin(), w.end(), ' ', '_');
+      }
+      // join string
+      std::stringstream s;
+      std::copy(words.begin(), words.end(), std::ostream_iterator<std::string>(s, " "));
+      val.def_val = s.str();
+      val.def_val = trim(val.def_val);
+    }
+  }
+
+  for (auto& m : dbc->msgs) {
+    m.sigs = signals[m.address];
+  }
+  for (auto& v : dbc->vals) {
+    v.sigs = signals[v.address];
+  }
+  return dbc;
 }
 
-extern "C" {
-  const DBC* dbc_lookup(const char* dbc_name) {
-    return dbc_lookup(std::string(dbc_name));
+void set_dbc_file_path() {
+  if (dbc_file_path.empty()) {
+    char *basedir = std::getenv("BASEDIR");
+    if (basedir != NULL) {
+      dbc_file_path = std::string(basedir) + "/opendbc";
+    } else {
+      dbc_file_path = DBC_FILE_PATH;
+    }
+  }
+}
+
+const DBC* dbc_lookup(const std::string& dbc_name) {
+  static std::mutex lock;
+  static std::map<std::string, DBC*> dbcs;
+
+  std::unique_lock lk(lock);
+  set_dbc_file_path();
+
+  auto it = dbcs.find(dbc_name);
```

Comment:
Since parsing DBCs is so quick, do we need this? Not sure how static `dbcs` works, if it caches anything or not